src/components/MagicCodeInput.js

```diff
@@ -0,0 +1,282 @@
+        PropTypes.shape({current: PropTypes.instanceOf(React.Component)}),
+    ]),
+
+    /** Error text to display */
+    errorText: PropTypes.string,
+
+    /** Specifies autocomplete hints for the system, so it can provide autofill */
+    autoComplete: PropTypes.oneOf(['sms-otp', 'one-time-code']).isRequired,
+
+    /* Should submit when the input is complete */
+    submitOnComplete: PropTypes.bool,
+
+    /** Id to use for this button */
+    nativeID: PropTypes.string,
+
+    /** Function to call when the input is changed  */
+    onChange: PropTypes.func,
+
+    /** Function to call when the input is submitted or fully complete */
+    onSubmit: PropTypes.func,
+};
+
+const defaultProps = {
+    value: undefined,
+    name: '',
+    autoFocus: true,
+    shouldDelayFocus: false,
+    forwardedRef: undefined,
+    errorText: '',
+    submitOnComplete: false,
+    nativeID: '',
+    onChange: () => {},
+    onSubmit: () => {},
+};
+
+/**
+ * Verifies if a string is a number.
+ *
+ * @param {string} value The string to check if it's numeric.
+ * @returns {boolean} True if the string is numeric, false otherwise.
+ */
+function isNumeric(value) {
+    if (typeof value !== 'string') { return false; }
+    return !Number.isNaN(value) && !Number.isNaN(parseFloat(value));
+}
+
+class MagicCodeInput extends React.PureComponent {
+    constructor(props) {
+        super(props);
+
+        this.inputNrArray = Array.from(Array(CONST.MAGIC_CODE_NUMBERS).keys());
+        this.inputRef = React.createRef(null);
+
+        this.state = {
+            input: '',
+            focusedIndex: 0,
+            editIndex: 0,
+            numbers: props.value ? this.decomposeString(props.value) : Array(CONST.MAGIC_CODE_NUMBERS).fill(''),
+        };
+
+        this.onChangeText = this.onChangeText.bind(this);
+        this.onKeyPress = this.onKeyPress.bind(this);
+    }
+
+    componentDidMount() {
+        if (!this.props.forwardedRef) {
+            return;
+        }
+        this.props.forwardedRef(this.inputRef);
+
+        if (!this.props.autoFocus) {
+            return;
+        }
+
+        if (this.props.shouldDelayFocus) {
+            this.focusTimeout = setTimeout(() => this.inputRef.focus(), CONST.ANIMATED_TRANSITION);
+            return;
+        }
+        this.inputRef.focus();
+    }
+
+    componentDidUpdate(prevProps) {
+        if (prevProps.value === this.props.value || this.props.value === this.composeToString(this.state.numbers)) {
+            return;
+        }
+
+        this.setState({
+            numbers: this.decomposeString(this.props.value),
+        });
+    }
+
+    /**
+     * Focuses on the input when it is pressed.
+     *
+     * @param {object} event The event passed by the input.
+     * @param {number} index The index of the input.
+     */
+    onFocus(event, index) {
+        event.preventDefault();
+        this.setState(prevState => ({
+            ...prevState, input: '', focusedIndex: index, editIndex: index,
```

Comment:
setState doesn't require to destruct previous state, setting only state for fields that we want to change is enough
```suggestion
            input: '', focusedIndex: index, editIndex: index,
```